test/e2e/storage/in_tree_volumes.go

```diff
@@ -0,0 +1,228 @@
+	drivers.InitCinderDriver,
+	drivers.InitGceDriver,
+	drivers.InitVSphereDriver,
+	drivers.InitAzureDriver,
+	drivers.InitAwsDriver,
+}
+
+// List of testSuites to be executed in below loop
+var testSuites = []func() testsuites.TestSuite{
+	testsuites.InitVolumesTestSuite,
+	testsuites.InitVolumeIOTestSuite,
+	testsuites.InitVolumeModeTestSuite,
+	testsuites.InitSubPathTestSuite,
+}
+
+// This executes testSuites for in-tree volumes.
+// It is done by following steps:
+// Loop over testDrivers:
+//   Loop over testSuites:
+//     - Get testPatterns from testSuites
+//     Loop over testPatterns:
+//       - Get test function to be executed
+//       BeforeEach test, prepare test by following steps:
+//          - Skip unsupported tests to avoid unnecessary resource initialization
+//          - Create Driver
+// 			- Create Volume for test unless dynamicPV test
+//          - Initialize test resource for testSuites
+//          - Setup test resource for driver and testpattern
+//          - Create test input to be passed to test function
+//       - Execute tests by calling test function with test input
+//       AfterEach test, cleanup everything by following steps:
+//          - Cleanup test resource
+// 			- Delete Volume for test unless dynamicPV test
+//          - Cleanup Driver
+var _ = utils.SIGDescribe("In-tree Volumes", func() {
+	f := framework.NewDefaultFramework("volumes")
+
+	var (
+		ns     *v1.Namespace
+		config framework.VolumeTestConfig
+
+		drv drivers.TestDriver
+		ts  testsuites.TestSuite
+		tr  testsuites.TestResource
+		ti  testsuites.TestInput
+		vt  testpatterns.TestVolType
+		tp  testpatterns.TestPattern
+
+		needsCleanup bool
+	)
+
+	BeforeEach(func() {
+		ns = f.Namespace
+		config = framework.VolumeTestConfig{
+			Namespace: ns.Name,
+			Prefix:    "volume",
+		}
+	})
+
+	for _, initDriver := range testDrivers {
+		curDriver := initDriver()
+
+		for _, testSuiteInit := range testSuites {
+			curTestSuite := testSuiteInit()
+			// Get testPatterns from testSuites
+			testPatterns := curTestSuite.GetTestPatterns()
+
+			for _, testPattern := range testPatterns {
+				curTestPattern := testPattern
+				ti = curTestSuite.CreateTestMetadata(curTestSuite, curTestPattern)
+				// Get test function to be executed
+				testFunc := curTestSuite.GetTestFunc(curTestPattern, curDriver)
+
+				Context(fmt.Sprintf(drivers.GetDriverNameWithFeatureTags(curDriver)), func() {
+					BeforeEach(func() {
+						// Cleanup for driver and resource won't be needed until they are created
+						needsCleanup = false
+						drv = curDriver
+						ts = curTestSuite
+						tp = curTestPattern
+						drivers.SetCommonDriverParameters(drv, f, config)
+
+						// Skip unsupported tests to avoid unnecessary resource initialization
+						skipUnsupportedTest(drv, ts, tp)
+
+						// Cleanup for driver and resource will be needed beyound here
+						needsCleanup = true
+
+						// Create driver
+						vt = tp.VolType
+						drv.CreateDriver(vt)
```

Comment:
Can Create/Cleanup Driver be called in the outermost loop?  I am thinking in the case of CSI, you may bring up the cluster and the driver once in the beginning, and then run all the test cases against it, and then bring down the driver/cluster at the end.

Reply:
I agree with you that it will be much cleaner. However, existing containerized volume test server, such as iSCSI and rbd, have a predefined volume inside. And, I think that existing tests are implemented in a way that volume, or the containerized volume test server, is expected to be initialized before each test. To move these test as they are, create/cleanup will be needed to be done for each test.

I think that we would be better improving this in the next step, after we achieve that all containerized volume test servers implement actual ```createVolume```/```deleteVolume```.

Reply:
Can the iscsi and rbd servers be moved to CreateVolume instead?